File: libs/base/serial.cpp
#include "pxtbase.h"

#ifdef CODAL_SERIAL
namespace pxt {
  class WSerial {
    public:
      CODAL_SERIAL serial;
      WSerial()
        : serial(LOOKUP_PIN(TX), LOOKUP_PIN(RX))
        {}
  }
  SINGLETON(WSerial);
}
#endif

namespace serial {
    void write(const char* buffer, int length) {
      #if CODAL_SERIAL
      getWSerial()->serial.write(text->data, text->length);
      #endif
    }
    /**
     * Write some text to the serial port.
     */
    //% help=serial/write-string
    //% weight=87 blockHidden=true
    //% blockId=serial_writestring block="serial|write string %text"
    void writeString(String text) {
      if (NULL == text) return;
      write(text->data, text->length);
    }

    /**
    * Send a buffer across the serial connection.
    */
    //% help=serial/write-buffer weight=6 blockHidden=true
    //% blockId=serial_writebuffer block="serial|write buffer %buffer"
    void writeBuffer(Buffer buffer) {
      if (NULL == buffer) return;
      write((char*)buffer->data, buffer->length);
    }
}

namespace console {
    /**
      Sends the console message through the TX, RX pins
      **/
    //% weight=1
    //% blockId=consolesendtoserial block="send console to serial"
    void sendConsoleToSerial() {
      #if CODAL_SERIAL
        setSendToUART(serial.write)
      #endif
    }
}

File: libs/base/serial.ts
/**
 * Reading and writing data over a serial connection.
 */
//% weight=2 color=#002050 icon="\uf287"
//% advanced=true
namespace serial {
    /**
     * Write a line of text to the serial port.
     * @param value to send over serial
     */
    //% weight=90
    //% help=serial/write-line blockGap=8 blockHidden=true
    //% blockId=serial_writeline block="serial|write line %text"
    export function writeLine(text: string): void {
        writeString(text + "\r\n");
    }

    /**
     * Write a number to the serial port.
     */
    //% help=serial/write-number
    //% weight=89 blockGap=8 blockHidden=true
    //% blockId=serial_writenumber block="serial|write number %value"
    export function writeNumber(value: number): void {
        writeString(value.toString());
    }

    /**
     * Write a name:value pair as a line of text to the serial port.
     * @param name name of the value stream, eg: "x"
     * @param value to write
     */
    //% weight=88 blockGap=8
    //% help=serial/write-value blockHidden=true
    //% blockId=serial_writevalue block="serial|write value %name|= %value"
    export function writeValue(name: string, value: number): void {
        const prefix = name ? name + ":" : "";
        serial.writeLine(prefix + value);
    }

}
